index.js

```diff
@@ -1,15 +1,56 @@
 const {promisify} = require('util')
 const fs = require('fs')
 const readFile = promisify(fs.readFile)
+const lstat = promisify(fs.lstat)
+const readdir = promisify(fs.readdir)
 const parse = require('json-parse-even-better-errors')
+
+const { resolve, dirname, join, relative } = require('path')
+
 const rpj = path => readFile(path, 'utf8')
-  .then(data => normalize(stripUnderscores(parse(data))))
+  .then(data => readBinDir(path, normalize(stripUnderscores(parse(data)))))
   .catch(er => {
     er.path = path
     throw er
   })
+
 const normalizePackageBin = require('npm-normalize-package-bin')
 
+// load the directories.bin folder as a 'bin' object
+const readBinDir = async (path, data) => {
+  if (data.bin)
+    return data
+
+  const m = data.directories && data.directories.bin
+  if (!m || typeof m !== 'string')
+    return data
+
+  // cut off any monkey business, like setting directories.bin
+  // to ../../../etc/passwd or /etc/passwd or something like that.
+  const root = dirname(path)
+  const dir = join('.', join('/', m))
+  data.bin = await walkBinDir(root, dir, {})
```

Comment:
so we if `directories.bin` points to a directory that contains directories, we flatten everything in all of those directories to a list of bin names? what if there's a collision? this feels like it should just be a readdir and not bother traversing subdirectories to me

Reply:
I agree.  But this is how it worked in v6 and before, so I thought it best to leave it as it is.  I'd suggest dropping support for this entirely in npm version 8 (or deprecating it in v8 and dropping in v9).